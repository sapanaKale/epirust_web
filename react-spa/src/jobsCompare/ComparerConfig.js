/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import Select from "react-select";
import React, {useState} from "react";
import {epiCurves} from "../common/constants";

export function ComparerConfig({selectedJobs, updateSelectedCurves, updateInterventions}) {
  const [selectedInterventions, updateSelectedInterventions] = useState({})

  const handleUpdateInterventions = (job, info) => {
    updateSelectedInterventions(prevState => ({...prevState, [job]: info}))
  }

  const handleSubmit = (e) => {
    e.preventDefault();

    const interventionsToAsk = []
    if (selectedInterventions.job1) interventionsToAsk.push(...selectedInterventions.job1)
    if (selectedInterventions.job2) interventionsToAsk.push(...selectedInterventions.job2)
    updateInterventions(interventionsToAsk)
  }

  return <div className="comparer-config">
    <div className="user-inputs">
      <div className="form-row">
        <h5><b><label htmlFor="series-filer">Select line-graphs:</label></b></h5>
        <Select
          options={epiCurves.map(e => ({value: e, label: e}))}
          isMulti
          inputId="series-filer"
          name="tags"
          aria-label="tags"
          onChange={(cs) => updateSelectedCurves(cs ? cs.map(c => c.value) : [])}
        />
      </div>
    </div>

    <form onSubmit={handleSubmit} className="user-inputs">
      <div className="form-row">
        <h5><b><label>Select interventions:</label></b></h5>
        <label htmlFor="job1-interventions-filter">Job 1:</label>
        <Select
          options={selectedJobs.job1.simulations.map(s => ({value: s._id, label: s._id}))}
          isMulti
          inputId="job1-interventions-filter"
          name="job1-interventions"
          aria-label="tags"
          onChange={(cs) => handleUpdateInterventions("job1", cs ? cs.map(c => c.value) : [])}
        />
      </div>
      <div className="form-row">
        <label htmlFor="job2-interventions-filter">Job 2:</label>
        <Select
          options={selectedJobs.job2.simulations.map(s => ({value: s._id, label: s._id}))}
          isMulti
          inputId="job2-interventions-filter"
          name="job2-interventions"
          aria-label="tags"
          onChange={(cs) => handleUpdateInterventions("job2", cs ? cs.map(c => c.value) : [])}
        />
      </div>
      <div className="form-row">
        <button className="btn btn-primary">Show Interventions</button>
      </div>
    </form>
  </div>
}
