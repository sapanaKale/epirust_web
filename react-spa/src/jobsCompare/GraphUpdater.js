/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import io from "socket.io-client";
import config from "../config";


export default class GraphUpdater {
  #updateBuffer
  #jobData1
  #jobData2
  #lastSent
  #stopAll

  constructor(updateBuffer, jobId1, jobId2) {
    this.#updateBuffer = updateBuffer;

    this.#jobData1 = new JobTimeSeriesData(jobId1)
    this.#jobData2 = new JobTimeSeriesData(jobId2)
    this.#lastSent = 0;
    this.#stopAll = false;
  }

  start() {
    this.#startJob(this.#jobData1);
    this.#startJob(this.#jobData2);
  }

  stop() {
    this.#jobData1.closeSocket();
    this.#jobData2.closeSocket();
  }

  #checkConsistencyAndSend = (sendLimit) => {
    const start = this.#lastSent + 1;
    const end = start + sendLimit - 1;
    const isConsistent = this.#checkConsistency(end)

    if (isConsistent) {
      const data = this.#buildData(start, end)
      if(!this.#stopAll) {
        this.#updateBuffer(data)
        this.#lastSent = end;
      }
    }
  }

  #startJob = (jobData) => {
    const socket = io(`${config.API_HOST}/counts`)
    jobData.setSocket(socket);
    socket.on('epidemicStats', (message) => {
      if ("simulation_ended" in message) {
        jobData.fetchFinished = true;
        if (this.#bothJobsFinished())
          this.#checkConsistencyAndSend(jobData.consumedCount() - this.#lastSent)
        socket.close()
      } else {
        jobData.insertData(message.hour, message)
        if (jobData.consumedCount() % config.RENDER_COUNT === 0) {
          this.#checkConsistencyAndSend(config.RENDER_COUNT)
        }
      }
    })
    socket.emit('get', {jobId: jobData.jobId});
  }

  #bothJobsFinished = () => {
    return this.#jobData1.fetchFinished && this.#jobData2.fetchFinished;
  }

  #checkConsistency = (till) => {
    const eitherJobsNotConsistent = this.#jobData1.consumedCount() < till || this.#jobData2.consumedCount() < till
    return !(!this.#bothJobsFinished() && eitherJobsNotConsistent);
  }

  #buildData = (from, till) => {
    const data = []
    for (let i = from; i <= till; i++) {
      data.push({'hour': i, 'job1': this.#jobData1.dataBuffer[i], 'job2': this.#jobData2.dataBuffer[i]})
    }
    return data;
  }
}

class JobTimeSeriesData {
  #consumedMessageCount
  #socket

  constructor(jobId) {
    this.jobId = jobId
    this.fetchFinished = false;
    this.dataBuffer = {}
    this.#consumedMessageCount = 0;
  }

  insertData = (hour, message) => {
    this.dataBuffer[hour] = message
    this.#consumedMessageCount += 1
  }

  consumedCount = () => this.#consumedMessageCount

  setSocket = (socket) => this.#socket = socket;

  closeSocket = () => this.#socket.close();
}
