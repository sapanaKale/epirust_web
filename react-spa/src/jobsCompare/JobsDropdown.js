/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import PropTypes from 'prop-types'

export default function JobsDropdown({jobs, label}) {
  const id = label.split(' ').join('').toLowerCase();
  return <>
    <label htmlFor={id}>{label}</label>
    <select name={id} className="form-control" id={id} required>
      <option disabled selected value=""> -- select a job -- </option>
      {jobs.map(j => <option key={j._id}>{j._id}</option>)}
    </select>
  </>
}

JobsDropdown.propTypes = {
  jobs: PropTypes.array.isRequired,
  label: PropTypes.string.isRequired,
}