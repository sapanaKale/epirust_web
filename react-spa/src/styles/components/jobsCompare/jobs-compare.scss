.jobs-compare-chart {
  display: flex;
  flex-flow: column;
  height: 80vh;
}

.comparer-config {
  padding-top: 20px;
}

.drawer-wrap {
  position: relative;

  .drawer {
    height: 100vh;
    width: 40%;

    z-index: 11;

    position: fixed;
    right:0;
    top: 0;

    transform: translateX(100%);
    transition: 0.2s transform;

    padding: 20px;

    &.open {
      background: white;
      transform: translateX(0);
      box-shadow: 0 10px 16px 0 rgba(0,0,0,0.2)
    }
  }
}