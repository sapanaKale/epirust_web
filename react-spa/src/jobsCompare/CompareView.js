/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {ComparerConfig} from "./ComparerConfig";
import Graph from "../time-series/LineGraph";
import React, {useEffect, useState} from "react";
import {epiCurves} from "../common/constants";
import {get} from "../common/apiCall";
import {parseAnnotations} from "../time-series/utils";
import GraphUpdater from "./GraphUpdater";
import SideBarContainer from "../common/SideBarContainer";

export function CompareView({selectedJobs}) {
  const [graphData, updateGraphData] = useState([]);
  const [selectedCurves, updateSelectedCurves] = useState([]);
  const [annotations, updateAnnotations] = useState([]);

  function clearGraphDataAndConfig() {
    updateGraphData([]);
    updateAnnotations([]);
    updateSelectedCurves([])
  }

  const updateBuffer = (buffer) => {
    updateGraphData(prevData => prevData.concat(buffer))
  };

  useEffect(() => {
    clearGraphDataAndConfig();
    const {job1, job2} = selectedJobs;
    const graphUpdater = new GraphUpdater(updateBuffer, job1._id, job2._id);
    graphUpdater.start();

    return () => graphUpdater.stop();

  }, [selectedJobs]);

  const onUpdateInterventions = (interventionsAskedFor) => {
    const series = selectedCurves.length > 0 ? `${selectedCurves[0]}_job1` : "susceptible_job1";

    function getInterventionsForSimulation(simulationId) {
      return get(`/simulations/${simulationId}/interventions`)
        .then(j => j.json())
        .then(res => {
          return res
            .map(i => parseAnnotations(i.interventions, i.hour))
            .reduce((acc, cur) => acc.concat(cur), [])
            .map(i => ({...i, series}))
        })
    }

    Promise.all(interventionsAskedFor.map(getInterventionsForSimulation))
      .then(multiParsedAnnotations => {
        const annotations = multiParsedAnnotations
          .reduce((acc, cur) => acc.concat(cur), [])
          .sort((a, b) => a.x < b.x);
        updateAnnotations(annotations)
      })
      .catch(err => {
        console.log("Failed to load all interventions", err)
      })
  };

  const handleSelectedCurvesUpdate = (curves) => {
    if (curves.length > 0) {
      updateAnnotations(prevState => prevState.map(x => ({...x, series: `${curves[0]}_job1`})))
    }
    updateSelectedCurves(curves)
  };

  return (
    <div>
      <SideBarContainer >
        <ComparerConfig
          selectedJobs={selectedJobs}
          updateSelectedCurves={handleSelectedCurvesUpdate}
          updateInterventions={onUpdateInterventions}
        />
      </SideBarContainer>
      <div className="jobs-compare-chart">
        {graphData.length > 0 &&
        <Graph
          dataBuffer={makeCSV(graphData)}
          dygraphsOptions={dygraphOptions()}
          visibility={visibility(selectedCurves)}
          annotations={annotations}
        />
        }
      </div>
    </div>
  );
}

const dygraphOptions = () => ({
  colors: ["#00a987", "#005b98", "#8900c1", "#e46b00", "#2c3e50",
    "#1abc9ccc", "#3498dbcc", "#9b59b6cc", "#e67e22cc", "#7f8c8d"],
  labels: ["hour",
    ...epiCurves.map(e => `${e}_job1`),
    ...epiCurves.map(e => `${e}_job2`),
  ]
});

const visibility = (curvesSelected) => {
  const curves = [...epiCurves.map(e => `${e}_job1`), ...epiCurves.map(e => `${e}_job2`)];
  if (curvesSelected.length > 0) {
    return curves.map(l => curvesSelected.some(cs => l.includes(cs)))
  }
  return curves.map(_ => true)
};

const rowToCsv = (row) => {
  if (row) return [[row.susceptible, row.susceptible_std], [row.infected, row.infected_std],
    [row.hospitalized, row.hospitalized_std], [row.recovered, row.recovered_std], [row.deceased, row.deceased_std]];
  return [[], [], [], [], []]
};

const makeCSV = (graphData) => {
  return graphData.map(d => ([d.hour, ...rowToCsv(d.job1), ...rowToCsv(d.job2)]))
};
