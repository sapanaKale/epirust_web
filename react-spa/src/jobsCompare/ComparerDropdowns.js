/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useState} from "react";
import JobsDropdown from "./JobsDropdown";
import PropTypes from "prop-types";

export default function ComparerDropdowns({jobs, onCompare}) {
  const [showError, updateShowError] = useState(false);

  const handleSubmit = (e) => {
    e.preventDefault();
    const formsData = new FormData(e.target);
    const selectedJobs = Object.fromEntries(formsData.entries())
    if (selectedJobs['job1'] === selectedJobs['job2']) {
      updateShowError(true)
    } else {
      onCompare(selectedJobs)
    }
  }

  return (
    <div className='comparer-dropdowns container'>
      <form className='row' onSubmit={handleSubmit}>
        <div className="col-5">
          <JobsDropdown jobs={jobs} label='Job 1'/>
        </div>
        <div className="col-5">
          <JobsDropdown jobs={jobs} label='Job 2'/>
        </div>
        <div className="col-2 margin-top-auto">
          <button className="btn btn-primary">Compare</button>
        </div>
        {showError && <div className="error-message alert-danger">Can't compare same jobs!</div>}
      </form>
    </div>
  );
}

ComparerDropdowns.propTypes = {
  jobs: PropTypes.array.isRequired,
  onCompare: PropTypes.func
}