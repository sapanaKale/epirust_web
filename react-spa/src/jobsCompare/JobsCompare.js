/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import ComparerDropdowns from "./ComparerDropdowns";
import {get} from "../common/apiCall";
import {LoadingComponent} from "../common/LoadingComponent";
import {LOADING_STATES} from "../common/constants";
import {reduceStatus} from "../jobs/JobTransformer";
import {CompareView} from "./CompareView";

export default function JobsCompare() {
  const [jobs, updateJobs] = useState([]);
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING);
  const [selectedJobs, updateSelectedJobs] = useState(null);

  useEffect(() => {
    get('/jobs')
      .then((res) => res.json())
      .then((jobsResponse) => {
        updateJobs(jobsResponse.map(reduceStatus).filter(j => j.status === 'finished'));
        updateLoadingState(LOADING_STATES.FINISHED)
      })
      .catch(() => updateLoadingState(LOADING_STATES.FAILED))
  }, []);

  const onCompare = (jobPair) => {
    updateSelectedJobs({
      job1: jobs.find(x => x._id === jobPair.job1),
      job2: jobs.find(x => x._id === jobPair.job2)
    })
  };

  return (
    <div>
      <LoadingComponent loadingState={loadingState}>
        <div className='row justify-content-md-center'>
          <div className='col-7'>
            <ComparerDropdowns jobs={jobs} onCompare={onCompare}/>
          </div>
        </div>
        {selectedJobs && <CompareView selectedJobs={selectedJobs}/>}
      </LoadingComponent>
    </div>
  )
}
