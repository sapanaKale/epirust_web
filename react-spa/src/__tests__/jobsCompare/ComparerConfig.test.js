/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import ShallowRenderer from 'react-test-renderer/shallow';
import {ComparerConfig} from "../../jobsCompare/ComparerConfig";
import React from "react";

describe('Comparer Config', () => {
  it('should render Select for time-series filer and form for selecting interventions', () => {
    const renderer = new ShallowRenderer()
    const selectedJobs = {
      job1: {simulations: [{_id:"4567uhg"}, {_id: "4567uyhghjn"}]},
      job2: {simulations: [{_id:"456776tgh"}]}
    }
    renderer.render(<ComparerConfig selectedJobs={selectedJobs}/>)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
  });
});
