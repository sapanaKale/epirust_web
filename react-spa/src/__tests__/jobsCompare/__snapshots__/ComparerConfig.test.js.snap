// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comparer Config should render Select for time-series filer and form for selecting interventions 1`] = `
<div
  className="comparer-config"
>
  <div
    className="user-inputs"
  >
    <div
      className="form-row"
    >
      <h5>
        <b>
          <label
            htmlFor="series-filer"
          >
            Select line-graphs:
          </label>
        </b>
      </h5>
      <StateManager
        aria-label="tags"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        inputId="series-filer"
        isMulti={true}
        name="tags"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "susceptible",
              "value": "susceptible",
            },
            Object {
              "label": "infected",
              "value": "infected",
            },
            Object {
              "label": "hospitalized",
              "value": "hospitalized",
            },
            Object {
              "label": "recovered",
              "value": "recovered",
            },
            Object {
              "label": "deceased",
              "value": "deceased",
            },
          ]
        }
      />
    </div>
  </div>
  <form
    className="user-inputs"
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <h5>
        <b>
          <label>
            Select interventions:
          </label>
        </b>
      </h5>
      <label
        htmlFor="job1-interventions-filter"
      >
        Job 1:
      </label>
      <StateManager
        aria-label="tags"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        inputId="job1-interventions-filter"
        isMulti={true}
        name="job1-interventions"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "4567uhg",
              "value": "4567uhg",
            },
            Object {
              "label": "4567uyhghjn",
              "value": "4567uyhghjn",
            },
          ]
        }
      />
    </div>
    <div
      className="form-row"
    >
      <label
        htmlFor="job2-interventions-filter"
      >
        Job 2:
      </label>
      <StateManager
        aria-label="tags"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        inputId="job2-interventions-filter"
        isMulti={true}
        name="job2-interventions"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "456776tgh",
              "value": "456776tgh",
            },
          ]
        }
      />
    </div>
    <div
      className="form-row"
    >
      <button
        className="btn btn-primary"
      >
        Show Interventions
      </button>
    </div>
  </form>
</div>
`;
